Update issue triaging docs (#794)
rust-lang/rust-forge: src/release/issue-triaging.md

## Motivation

The [`rust-lang/rust`] repository has thousands of issues and hundreds of people
working on it. It is impossible for all people to check and solve issues. The
goals of triaging are to:

- Improve **discoverability**:
    - Categorize which teams an issue may concern.
    - Make sure regressions are properly tagged and tracked.
    - Connect issues to the relevant people (especially relevant teams).
- First-pass **coarse classification** of the issue nature:
    - Is this issue a bug, or is this discussion, or is this a user-side
      problem?
- Identify if the issue is **sufficiently actionable**:
    - Does the report need further info, like reproduction steps, a reproducer,
      or build/execution environment info?
    - Does the reproducer need minimization?
- Make it easier for contributors to **filter** issues by applying suitable
  labels.

In practice, it is unrealistic for all issues to be solved quickly and found by
right people. Through applications of labels we make the issue tracker more
searchable for future reference, so that people in the future have an easier
time finding related issues or issues they are interested in working on.

> Triaging can be done by **everyone**, no matter your permissions. We encourage

Author: That is true. Maybe "assisted"?

> Triaging can be done by **everyone**, no matter your permissions. We encourage

Reviewer: some actions need permissions, but a lot don't, so I think the statement is accurate. it would be useful to mention what to do about actions that do

> Triaging can be done by **everyone**, no matter your permissions. We encourage

Author: Remark (for myself): `@rustbot transfer`, closing, adding write-access-gated labels

> Triaging can be done by **everyone**, no matter your permissions. We encourage

Author: Further clarified this to say something like:

---

Triaging can be done by **everyone**, no matter if you have write access to [`rust-lang/rust`]. We encourage everyone to help here, as triaging work is highly parallelizable and easy to get started with.

Some actions that require write access to [`rust-lang/rust`] include:

- Using `@rustbot transfer` to transfer the issue to another repository under `rust-lang` GitHub organization.
- Closing the issue.
- Using `@rustbot` to apply certain labels that require write access (see "Applying and removing labels" section below).

If you don't have write access but you assess that these actions should be done, you can just open a new topic in [`t-release/triage`] zulip channel, or add a comment linking to the issue in an existing topic in the [`t-release/triage`] channel.

everyone to help here, as triaging work is highly parallelizable and easy to get
started with.

## Initial triaging

When an issue is opened, it usually receives the `needs-triage` label. This

> ensures that every issue gets an initial look and that no issue is never looked

Author: Yeah.

> ensures that every issue gets an initial look and that no issue is never looked

Author: Fixed

> at.

Author: Fixed

`needs-triage` is an initial checkpoint. The effort needed to get an issue past
the label should be small.

To do the initial triage and remove the `needs-triage` label, the following
conditions should be fulfilled/considered. It's okay if not all of these are
always considered; treat it as a guideline, not a hard checklist. It is also not

> exhaustive.

Reviewer 2: Can we remove reformatting changes from this patch or is it too inconvenient? I'd also probably stick to the current formatting choice for this document (which IIUC has no chars/line limit)

Author: I'll undo the reformatting, it was just more convenient when editing, and go with the no-column limit / sentence wrapping. We don't have any consistent formatting choices for docs in this repo, lol.

Author: Undid formatting changes, now all should be unbounded in column width.

- The issue should make sense, that is, it should present a problem.
    - For example, if an issue is a question about Rust in general, the issue

>       should be closed and the user redirected to [URLO]/[Discord]. You can of
>       course answer the question too, but make sure to mention that the user
>       should go to [URLO]/[Discord] next time.

Reviewer 3: ```suggestion
      should be closed and the user redirected to [URLO]/[Discord]. You can, of
      course, answer the question too, but make sure to mention that the user
      should go to [URLO]/[Discord] next time.
```

Author: Fixed

      the most important, because team members will usually rely on those as a
      baseline to filter what concerns the team (e.g. during triage meetings).
- If the issue contains no reproduction but needs one (when in doubt, it needs
  one), ask for one and add the `S-needs-repro` label.
- If the issue lacks information (e.g. system information, the exact invocation
  used), then request that the reporter provide the information that is
  important to reproduce the problem. Add the `S-needs-info` label.
- The issue tracker is the wrong place for some kinds of feature requests.
  Redirect the author to proper channels:
    - Standard library API requests should follow [libs-api
      processes](https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html).
    - Language changes should be redirected to [IRLO] or T-lang zulip channel
      ([t-lang](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang)).
- If the issue could benefit from bisecting the regression (when in doubt, it
  can), add `E-needs-bisection` (or do the bisection yourself).
    - If there is a bisection and the bisection outcome is convincing, add

>       `S-has-bisection`.

Author: Added a link

- If the reported example is large and/or complex (e.g. a lot of code, has a lot
  of external dependencies, has unrelated warnings or errors that obfuscates the
  actual problem), add `E-needs-mcve` to indicate the need to minimize the
  example.
    - If there is a Minimal, Complete and Verifiable Example (MCVE), then tag
      the issue with `S-has-mcve`. Double-check that the minimization is valid,
      e.g. the MCVE is for the original problem and has not drifted to a
      different problem.
- Is the issue a regression? If so, apply the `regression-untriaged` label (or

>   figure out what regression it is exactly).

Author: I don't recall exactly if we do... @apiraino do you happen to know what we consider a regression? Because also T-compiler vs other teams might use this differently...

>   figure out what regression it is exactly).

Reviewer: generally it's a regression if it worked but no longer works.

>   figure out what regression it is exactly).

Author: True ~~though sometimes that can be a progression~~

>   figure out what regression it is exactly).

Author: Added clarification

- If you happen to know people who this issue is relevant to, ping them.
    - For example, write `cc @ThatPerson` if `ThatPerson` has been working a lot
      on the feature in question recently.
- Does this issue require nightly? Add `requires-nightly`.
- Does this issue require incomplete or internal features? Add
  `requires-{incomplete,internal}-features`.
- Check if this issue is a duplicate of earlier-reported issues.
    - If you are certain this is a duplicate, close this issue as a duplicate of
      the earlier issue. Make sure this is obvious in the backlink of the
      earlier issue, or explicitly link to the duplicate issue.
    - If you are not sure, you can still leave a comment to indicate the other

>       issue is possibly a duplicate, similar, or related.

Author: ... Yeah, good point.

>       issue is possibly a duplicate, similar, or related.

Author: Moved higher up

> For applying and removing labels, unprivileged users can use **@rustbot** to add

Reviewer 2: why does "unprivileged" (or whatever adjective we prefer) need to be specified? Isn't `@rustbot` just for everyone to use (at least for some commands)? Or do I remember wrong?

> For applying and removing labels, unprivileged users can use **@rustbot** to add

Author: You have to be a member on some team in order to use certain labels, I believe: see https://github.com/rust-lang/rust/blob/13738b0abeeb6b7b8b52cc16aa8ece267b58f017/triagebot.toml#L5-L32 for an allowlist of labels that can be applied without the priviledges. E.g. I believe you can't apply `P-*` labels if you don't have write access to r-l/r.

> For applying and removing labels, unprivileged users can use **@rustbot** to add

Reviewer: I don't think people with write access should use rustbot because it sends our notifications to everyone (and has a worse UI), the normal GitHub tools are better. rustbot is just a workaround

> For applying and removing labels, unprivileged users can use **@rustbot** to add

Reviewer: also if we want to reword the "unprivileged", let's change it to "users without write access"

> For applying and removing labels, unprivileged users can use **@rustbot** to add

Author: Reworded "unpriviledged" -> "without write access". Also linked to the `[relabel]` section in triagebot.toml.

or remove [the labels allowed by the `triagebot.toml`
configuration](https://github.com/rust-lang/rust/blob/master/triagebot.toml).

For example, `@rustbot label +T-compiler +C-bug +A-linkage +O-macos
-needs-triage`.

To see a list of all labels, check out the "labels" page next to the search bar
in the issue tracker.

### Labels

There are many different labels that can be applied to issues.

- `needs-triage`: Signals that an issue is new and needs initial triage.
- [`T-*`]: Specifies the team or teams that this issue is relevant to. For
      example `T-compiler`, `T-types` or `T-libs`.
- [`WG-*`]: Specifies the working groups that this issue is relevant to, for
      example `WG-debugging`.
- [`PG-*`]: Specifies the project groups that this issue is relevant to, for
      example the `PG-exploit-mitigations`.
- [`C-*`]: Specifies the category of the label, for example a bug, tracking
      issue or discussion.

>     - `A-diagnostics`: issues usually don't have any `C-*` label.

Reviewer 3: Is it something should be mentioned under [`A-*`] instead

Author: Probably. But also `A-diagnostics` is a bit weird because of `A-diagnostics` issue template don't apply `C-bug` by default (intentionally).

Author: Moved under `A-*`

- [`A-*`]: The areas that the issue is relevant to, for example `A-linkage`,
      `A-patterns`, `A-diagnostics`. This is particularly helpful for filtering
      issues.
- [`D-*`]: Labels for diagnostic issues.
    - `D-diagnostic-infra`: This issue is about the diagnostics infrastructure
      itself.
- [`L-*`]: When the issue concerns a specific lint.
    - `L-false-positive` if the lint fires on a case that it should not have
      fired on.
    - `L-false-negative` if the lint misses a case where it should have fired
      on.
- [`F-*`]: When the issue concerns a specific (usually unstable, usually
      language) feature.
- [`-Z*`]: When the issue concerns a specific unstable `-Z` compiler flag.
- `requires-nightly`: This issue only affects the nightly channel compiler, and
  does not affect the beta or stable channel compiler.
- `requires-{incomplete,internal}-features`: This issue requires an incomplete
  or internal feature. The latter often means that the issue should be closed in
  accordance with compiler [MCP
  620](https://github.com/rust-lang/compiler-team/issues/620).
- [`regression-*`]: Labels for tracking issues that are regressions.

>     - `regression-untriaged`: unclear what kind of regression and its nature,
>       indicates further regression triaging is necessary.
>     - `regression-from-stable-to-stable`: something that regressed from an older
>       stable release, which has already reached the latest stable release or
>       even earlier stable releases.
>     - `regression-from-stable-to-{beta,nightly}`: something that regressed from
>       the a stable release to beta or nightly channel.

Reviewer 4: ```suggestion
    - `regression-untriaged`: The nature and kind of regression is unclear;
      indicates further regression triaging is necessary.
    - `regression-from-stable-to-stable`: Something that regressed from an older
      stable release, which has already reached the latest stable release or
      even earlier stable releases.
    - `regression-from-stable-to-{beta,nightly}`: Something that regressed from
      a stable release to beta or nightly channel.
```

Author: Fixed

> - [`I-*`]: Different labels about the nature[^2] of a bug.

Reviewer 4: ```suggestion
- [`I-*`]: Different labels about the nature[^2] of an issue.
```

    - `I-ICE`: Internal Compiler Error.

>     - `I-prioritize`: indicates that the issue should be additionally triaged by
>       the [T-compiler prioritization working group
>       (WG-priority)](../compiler/prioritization.md) to assign a priority `P-*`
>       if applicable to indicate the urgency.
>     - `I-heavy`: heavy code (binary size).
>     - `I-slow`: slow run-time performance.
>     - `I-hang`: the compilation fails to complete after a long time.
>     - `I-crash`: the compiler or generated code crashes but does not manifest as
>       an ICE, e.g. SIGSEGV or access violations.
>     - `I-unsound`: library, compiler, type system unsoundness.
>     - `I-compilemem`: excessive memory usage during compilation.
>     - `I-compiletime`: slow compilation time. 
>     - `I-{team}-nominated`: issue is nominated for discussion by `{team}`. E.g.
>       `I-compiler-nominated`.

Reviewer 4: ```suggestion
    - `I-prioritize`: Indicates that the issue should be additionally triaged by
      the [T-compiler prioritization working group
      (WG-priority)](../compiler/prioritization.md) to assign a priority `P-*`
      if applicable to indicate the urgency.
    - `I-heavy`: Heavy code (binary size).
    - `I-slow`: Slow run-time performance.
    - `I-hang`: The compilation fails to complete after a long time.
    - `I-crash`: The compiler or generated code crashes but does not manifest as
      an ICE, e.g. SIGSEGV or access violations.
    - `I-unsound`: Library, compiler, type system unsoundness.
    - `I-compilemem`: Excessive memory usage during compilation.
    - `I-compiletime`: Slow compilation time. 
    - `I-{team}-nominated`: Issue is nominated for discussion by `{team}`. E.g.
      `I-compiler-nominated`.
    - `I-lang-easy-decision`: The decision needed by lang is conjectured (by the
      person applying the label) to be easy or perfunctory.
```

Capitalized and added documentation of `I-lang-easy-decision`.

- [`P-*`]: Priority labels. Applied using the [compiler prioritization
  procedure](../compiler/prioritization.md).
- [`E-*`]: Calls for participation[^3], for example to minimize an issue.

>     - `E-mentor`: A mentor is available to help with the issue, which makes for
>       good first issues. Make sure that the intended mentor is actually willing
>       and available to mentor.

Reviewer 4: ```suggestion
    - `E-mentor`: A mentor is available to help with the issue, which makes for
      a good first issue. Make sure that the intended mentor is actually willing
      and available to mentor.
```

>     - `E-needs-mcve`: This issue has a reproduction, but it is not minimal, it
>       should be minimized.

Reviewer 4: ```suggestion
    - `E-needs-mcve`: This issue has a reproduction, but it is not minimal; it
      should be minimized.
```

Author: Fixed.

    - `E-needs-bisection`: This issue needs a bisection, for example using
      [cargo-bisect-rustc](https://github.com/rust-lang/cargo-bisect-rustc).
    - `E-needs-investigation`: This issue needs further investigation to
      determine root causes and the nature of the issue.
    - `E-needs-design`: This issue will require some substantial design effort
      (exploration, prototyping, discussions, etc.).
    - `E-needs-test`: The issue has been fixed, but no test has been added for
      it. After someone adds a test, it can be closed.
    - `E-{easy,medium,hard}`: Someone has estimated how hard the issue is to
      fix. This can help with finding good first issues, but is [bound to be
      inaccurate](https://en.wikipedia.org/wiki/Curse_of_knowledge).
- [`S-*`]: The status of an issue, for example `S-needs-repro` or
  `S-needs-info`.

>     - `S-needs-info`: needs more information from issue reporter.
>     - `S-needs-repro`: needs reproducer (code sample, repro steps, etc.).
>     - `S-bug-has-test`: issue has a known-bug test in [`rust-lang/rust`]
>       `tests/crashes` test suite or elsewhere.
>     - `S-has-bisection`: a bisection has been performed and the bisection result
>       is convincing.
>     - `S-waiting-on-LLVM`: waiting on upstream LLVM PR or fix.
>     - `S-tracking-forever`: the tracking issue is never intended to be closed.

Reviewer 4: ```suggestion
    - `S-needs-info`: Needs more information from issue reporter.
    - `S-needs-repro`: Needs reproducer (code sample, repro steps, etc.).
    - `S-bug-has-test`: Issue has a known-bug test in [`rust-lang/rust`]
      `tests/crashes` test suite or elsewhere.
    - `S-has-bisection`: A bisection has been performed and the bisection result
      is convincing.
    - `S-waiting-on-LLVM`: Waiting on upstream LLVM PR or fix.
    - `S-tracking-forever`: The tracking issue is never intended to be closed.
```

- Remove `needs-triage` if the relnotes wording has been confirmed/changed by
  the PR author, reviewer or a relevant team member.

### Internal Compiler Error (ICE) issue triage

For issues that include an Internal Compiler Error (ICE):

- Make sure the issue is tagged with `I-ICE`, `T-compiler` and `C-bug`.
- Check that the issue is actually an ICE, and not more accurately described
  with `I-crash` or `I-hang`.
- If it is an older (like latest stable) version of rust, ask:
    - Does the ICE reproduce on beta and latest nightly.
    - If older stable, ask if the ICE reproduces on latest stable.
- Check for duplicates, but don't close as duplicate unless you're sure they
  represent the same underlying issue. Prefer simply linking to the issue as
  possibly related/duplicate.
    - A good way is to search the ICE message. Be careful to not close as
      duplicate when the same ICE message may have different root causes.
- If it does not have a reproduction, comment asking for one and add
 ` S-needs-repro`.
    - If there isn't one for around a month it should generally be closed.
- If the reproduction is not minimal, add `E-needs-mcve` or create a MCVE

>   yourself. See the Initial triaging section for more details on MCVE procedure.

Reviewer 4: ```suggestion
  yourself. See the "Initial triaging" section for more details on MCVE procedure.
```

Author: Fixed

- Add `A-*` labels based on the code that causes the issue (check backtraces!),

>   and the nature of the repro (eg. if the repro is a weird trait impl or the

Reviewer 4: ```suggestion
  and the nature of the repro (e.g. if the repro is a weird trait impl or the
```

Author: Fixed

>   backtrace points to `rustc_trait_selection`, add `A-traits`)

Reviewer 4: ```suggestion
  backtrace points to `rustc_trait_selection`, add `A-traits`).
```

Author: Fixed

- Add `T-*`, `WG-*`, `PG-*`, `F-*`, `requires-*`, and `regression-*` labels as
  appropriate.
    - Usually the ICE should be tagged `T-compiler`. If the issue concerns the
      type system (e.g. trait solver), also tag the issue with `T-types`.

## Further triaging

For issues that have been through the initial triaging step (that is, don't have
the `needs-triage` label anymore), there are usually still things that can be
improved. There are often many more labels that could be applied (using rustbot
again if you don't have privileges).

Additionally, old (there is no clear definition of old yet, but something on the
order of months) `S-needs-repro` issues can be closed if there is no way to make
progress without a reproduction. This requires privileges, but if you don't have
them, you can just link the issue on Zulip (for example in `t-release/triage` or
`general`) and someone with privileges can close it for you.

> Another useful thing to do is go through `E-needs-mcve` and `E-needs-bisection`
> issues and creating minimizations or bisecting the issue (using

Reviewer 4: ```suggestion
Another useful thing to do is to go through `E-needs-mcve` and
`E-needs-bisection` issues and create minimizations or bisections (using
```

Author: Fixed

# Issue triaging

> > This page was last updated on: 2025-01-06.

Reviewer 5: Why is this annotation useful? It seems prone to rot, and folks who care can check the git history right?

Author: The intention is for a quick remark of when this page was last revised.

Author: Dropped

the label should be small.

To do the initial triage and remove the `needs-triage` label, the following
conditions should be fulfilled/considered. It's okay if not all of these are
always considered; treat it as a guideline, not a hard checklist. It is also not
exhaustive.

- The issue should make sense, that is, it should present a problem.
    - For example, if an issue is a question about Rust in general, the issue
      should be closed and the user redirected to [URLO]/[Discord]. You can of
      course answer the question too, but make sure to mention that the user
      should go to [URLO]/[Discord] next time.
- Add appropriate labels ([Labels](#labels)).
    - Specifically, team labels (`T-*`) and issue category labels (`C-*`) are
      the most important, because team members will usually rely on those as a
      baseline to filter what concerns the team (e.g. during triage meetings).
- If the issue contains no reproduction but needs one (when in doubt, it needs
  one), ask for one and add the `S-needs-repro` label.
- If the issue lacks information (e.g. system information, the exact invocation

>   used), then request that the reporter provide important that is important to

Reviewer 4: ```suggestion
  used), then request that the reporter provide the information that is important to
```

Author: Fixed

  produce the problem. Add the `S-needs-info` label.
- The issue tracker is the wrong place for some kinds of feature requests.
  Redirect the author to proper channels:
    - Standard library API requests should follow [libs-api
      processes](https://std-dev-guide.rust-lang.org/development/feature-lifecycle.html).
    - Language changes should be redirected to [IRLO] or T-lang zulip channel
      ([t-lang](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang)).
- If the issue could benefit from bisecting the regression (when in doubt, it
  can), add `E-needs-bisection` (or do the bisection yourself).
    - If there is a bisection and the bisection outcome is convincing, add
      `S-has-bisection`.
- If the reported example is large and/or complex (e.g. a lot of code, has a lot
  of external dependencies, has unrelated warnings or errors that obfuscates the
  actual problem), add `E-needs-mcve` to indicate the need to minimize the
  example.
    - If there is a Minimal, Complete and Verifiable Example (MCVE), then tag
      the issue with `S-has-mcve`. Double-check that the minimization is valid,

>       e.g. the MCVE is for the original problem and not drifted to a different

Reviewer 4: ```suggestion
      e.g. the MCVE is for the original problem and has not drifted to a different
```

Author: Fixed

  problem.
- Is the issue a regression? Apply the `regression-untriaged` label (or figure

>   out what regression it is exactly)

Reviewer 4: ```suggestion
  out what regression it is exactly).
```

Author: Fixed

    - If you are not sure, you can still leave a comment to indicate the other
      issue is possibly a duplicate, similar, or related.

For applying and removing labels, unprivileged users can use **@rustbot** to add
or remove [the labels allowed by the `triagebot.toml`
configuration](https://github.com/rust-lang/rust/blob/master/triagebot.toml).

For example, `@rustbot label +T-compiler +C-bug +A-linkage +O-macos
-needs-triage`.

To see a list of all labels, check out the "labels" page next to the search bar
in the issue tracker.

### Labels

There are many different labels that can be applied to issues.

- `needs-triage`: Signals that an issue is new and needs initial triage.
- [`T-*`]: Specifies the team or teams that this issue is relevant to. For
      example `T-compiler`, `T-types` or `T-libs`.
- [`WG-*`]: Specifies the working groups that this issue is relevant to, for
      example `WG-debugging`.
- [`PG-*`]: Specifies the project groups that this issue is relevant to, for
      example the `PG-exploit-mitigations`.
- [`C-*`]: Specifies the category of the label, for example a bug, tracking
      issue or discussion.

>     - `A-diagnostics`: issues usually don't have any `C-*` label.

Reviewer 4: It's perhaps surprising for an `A-*` item to be the first in a list of `C-*` items.  Also, in general, this line reads rather ambiguously.

Doing a quick [search](https://github.com/rust-lang/rust/issues?q=is%3Aissue%20state%3Aopen%20label%3AA-diagnostics&page=1), quite a lot of issues labels `A-diagnostics` do also have `C-*` labels.

Author: It's a bit weird because we don't do this consistently, at all. Issue templates for diagnostics issues don't add `C-bug` by default, but you can manually add it. Also some issues involve diagnostics but are actually bugs e.g. ICEs. I'll add some stuff to clarify this.

Author: Moved under `A-*` and described this behavior due to new diagnostics issue template.

>     - `C-optimization`: for missed compiler optimizations.
>     - `C-defective-hardware`: for hardware bugs that are beyond our control.

Reviewer 4: ```suggestion
    - `C-optimization`: For missed compiler optimizations.
    - `C-defective-hardware`: For hardware bugs that are beyond our control.
```

Elsewhere, the leading letter is capitalized even when it's a sentence fragment, so probably it's just worth being consistent in that direction.

Author: Fixed

>     - `C-gub`: if the problem is due to user error, e.g. Undefined Behavior.

Author: Fixed

>     - `C-external-bug`: for software bugs that affect us but we don't have direct

Reviewer 4: ```suggestion
    - `C-external-bug`: For software bugs that affect us but we don't have direct
```

      control over, but is worth tracking from our side.
- [`O-*`]: For target-specific issues, specifies the compile target[^1] or
      compile target family (most notably the platform, i.e., the architecture
      or operating system). For example `O-macos`, `O-aarch64`, `O-windows`,
      `O-windows-msvc`.

>     - Where possible, specify the most specific target-category where possible.

Reviewer 6: ```suggestion
    - Where possible, specify the most specific target-category.
```

Author: (Fixed, as this is possible :3)

      if an issue affects `*-windows-gnu` but not `*-windows-msvc`, prefer
      `O-windows-gnu` over the generic `O-windows`.
- [`A-*`]: The areas that the issue is relevant to, for example `A-linkage`,
      `A-patterns`, `A-diagnostics`. This is particularly helpful for filtering
      issues.
- [`D-*`]: Labels for diagnostic issues.
    - `D-diagnostic-infra`: This issue is about the diagnostics infrastructure
      itself.
- [`L-*`]: When the issue concerns a specific lint.
    - `L-false-positive` if the lint fires on a case that it should not have
      fired on.

>     - `L-false-negative` if the lint misses a case where it should have fired
>       on.

Reviewer 4: ```suggestion
    - `L-false-negative` if the lint misses a case where it should have fired.
```

Author: Fixed

- [`F-*`]: When the issue concerns a specific (usually unstable, usually
      language) feature.
- [`-Z*`]: When the issue concerns a specific unstable `-Z` compiler flag.
- `requires-nightly`: This issue only affects the nightly channel compiler, and
  does not affect the beta or stable channel compiler.
- `requires-{incomplete,internal}-features`: This issue requires an incomplete
  or internal feature. The latter often means that the issue should be closed in
  accordance with compiler [MCP
  620](https://github.com/rust-lang/compiler-team/issues/620).
- [`regression-*`]: Labels for tracking issues that are regressions.
    - `regression-untriaged`: unclear what kind of regression and its nature,
      indicates further regression triaging is necessary.
    - `regression-from-stable-to-stable`: something that regressed from an older
      stable release, which has already reached the latest stable release or
      even earlier stable releases.
    - `regression-from-stable-to-{beta,nightly}`: something that regressed from
      the a stable release to beta or nightly channel.
- [`I-*`]: Different labels about the nature[^2] of a bug.

>     - `I-ICE`: Internal Compiler Error.

Reviewer 4: ```suggestion
    - `I-ICE`: Internal compiler error.
```

Author: Fixed

      and available to mentor.
    - `E-needs-mcve`: This issue has a reproduction, but it is not minimal, it
      should be minimized.
    - `E-needs-bisection`: This issue needs a bisection, for example using
      [cargo-bisect-rustc](https://github.com/rust-lang/cargo-bisect-rustc).
    - `E-needs-investigation`: This issue needs further investigation to
      determine root causes and the nature of the issue.
    - `E-needs-design`: This issue will require some substantial design effort
      (exploration, prototyping, discussions, etc.).
    - `E-needs-test`: The issue has been fixed, but no test has been added for
      it. After someone adds a test, it can be closed.
    - `E-{easy,medium,hard}`: Someone has estimated how hard the issue is to
      fix. This can help with finding good first issues, but is [bound to be
      inaccurate](https://en.wikipedia.org/wiki/Curse_of_knowledge).
- [`S-*`]: The status of an issue, for example `S-needs-repro` or
  `S-needs-info`.
    - `S-needs-info`: needs more information from issue reporter.
    - `S-needs-repro`: needs reproducer (code sample, repro steps, etc.).

>     - `S-bug-has-test`: issue has a known-bug test in [`rust-lang/rust`]

Reviewer 6: Is the term *known-bug* explained/defined anywhere in the forge already? If not, we might want to rephrase *known-bug test* as *test marked with compiletest directive `//@ known-bug`* or something similar.

Author: So I didn't actually spell out the directive, because some known-bugs are actually just normal this-doesnt-compile `//@xxx-fail` tests, if you know what I mean

- Check for duplicates, but don't close as duplicate unless you're sure they
  represent the same underlying issue. Prefer simply linking to the issue as
  possibly related/duplicate.
    - A good way is to search the ICE message. Be careful to not close as
      duplicate when the same ICE message may have different root causes.
- If it does not have a reproduction, comment asking for one and add
 ` S-needs-repro`.
    - If there isn't one for around a month it should generally be closed.
- If the reproduction is not minimal, add `E-needs-mcve` or create a MCVE
  yourself. See the Initial triaging section for more details on MCVE procedure.
- Add `A-*` labels based on the code that causes the issue (check backtraces!),
  and the nature of the repro (eg. if the repro is a weird trait impl or the
  backtrace points to `rustc_trait_selection`, add `A-traits`)
- Add `T-*`, `WG-*`, `PG-*`, `F-*`, `requires-*`, and `regression-*` labels as
  appropriate.
    - Usually the ICE should be tagged `T-compiler`. If the issue concerns the
      type system (e.g. trait solver), also tag the issue with `T-types`.

## Further triaging

> See also section [Issue Triage](https://rustc-dev-guide.rust-lang.org/contributing.html#issue-triage) in the Rust Compiler Development Guide.

Author: Deleted this, I wanted to only keep label docs and stuff in one place, and forge docs seemed like a better fix. I'll send a r-d-g side PR to backlink to this page.

Author: https://github.com/rust-lang/rustc-dev-guide/pull/2194
